system/security/cryptohelper/pke.cpp

```diff
@@ -0,0 +1,245 @@
+/*##############################################################################
+
+    HPCC SYSTEMS software Copyright (C) 2018 HPCC Systems®.
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+############################################################################## */
+
+#if defined(_USE_OPENSSL) && !defined(_WIN32)
+
+#include "jliball.hpp"
+#include <openssl/pem.h>
+#include <openssl/err.h>
+#include <openssl/evp.h>
+
+#include "cryptocommon.hpp"
+#include "pke.hpp"
+
+namespace cryptohelper
+{
+
+void CLoadedKey::loadKeyBio(const char *keyMem)
+{
+    keyBio.set(BIO_new_mem_buf(keyMem, -1));
+    if (!keyBio)
+        throwEVPException(0, "loadKeyBio: Failed to create bio for key");
+}
+
+void CLoadedKey::loadKeyFileToMem(MemoryBuffer &keyMb, const char *keyFile)
+{
+    OwnedIFile iFile = createIFile(keyFile);
+    OwnedIFileIO iFileIO = iFile->open(IFOread);
+    if (!iFileIO)
+        throwEVPException(0, "loadKeyFileToMem: failed to open key: %s", keyFile);
+    size32_t sz = read(iFileIO, 0, iFile->size(), keyMb);
+    iFileIO.clear();
+}
+
+void CLoadedKey::finalize(RSA *rsaKey, const char *_keyName)
+{
+    if (!rsaKey)
+        throwEVPException(0, "Failed to create key: %s", _keyName);
+    key.set(EVP_PKEY_new());
+    EVP_PKEY_set1_RSA(key, rsaKey);
+    keyName.set(_keyName);
+}
+
+class CLoadedPublicKeyFromFile : public CLoadedKey
+{
+public:
+    CLoadedPublicKeyFromFile(const char *keyFile, const char *passPhrase)
+    {
+        MemoryBuffer keyMb;
+        loadKeyFileToMem(keyMb, keyFile);
```

Comment:
Not a fan of constructors that can throw

Reply:
exceptions in constructors are usually fine and in generally recommended of signalling failure during object creation.
You may be thinking of the problems associated with exceptions in destructors, which are problematic because of the way exception handling and stack unwinding happens.

Leaving as is.